File: shared-libs/node/src/logger.js
'use strict';

/**
 * Structured JSON logger with correlation ID support.
 * Wraps Winston with automatic correlationId from AsyncLocalStorage context.
 */

const { AsyncLocalStorage } = require('async_hooks');
const winston = require('winston');

const asyncLocalStorage = new AsyncLocalStorage();

const logger = winston.createLogger({
  level: process.env.LOG_LEVEL || 'info',
  format: winston.format.combine(
    winston.format.timestamp(),
    winston.format.json()
  ),
  transports: [new winston.transports.Console()],
});

function enrichWithContext(meta) {
  const store = asyncLocalStorage.getStore();
  if (store && store.correlationId) {
    return { correlationId: store.correlationId, ...meta };
  }
  return meta;
}

const log = {
  info:  (msg, meta = {}) => logger.info(msg, enrichWithContext(meta)),
  warn:  (msg, meta = {}) => logger.warn(msg, enrichWithContext(meta)),
  error: (msg, meta = {}) => logger.error(msg, enrichWithContext(meta)),
  debug: (msg, meta = {}) => logger.debug(msg, enrichWithContext(meta)),
  /** Run a function within a correlation-ID context. */
  withCorrelationId: (correlationId, fn) => asyncLocalStorage.run({ correlationId }, fn),
};

module.exports = { log, logger };

File: shared-libs/go/utils/logger/logger.go
// Package logger provides a structured JSON logger using slog.
package logger

import (
	"context"
	"log/slog"
	"os"
	"strings"
)

type ctxKey string

const correlationIDKey ctxKey = "correlationId"

var defaultLogger = newLogger()

func newLogger() *slog.Logger {
	level := slog.LevelInfo
	if l := os.Getenv("LOG_LEVEL"); l != "" {
		switch strings.ToLower(l) {
		case "debug":
			level = slog.LevelDebug
		case "warn", "warning":
			level = slog.LevelWarn
		case "error":
			level = slog.LevelError
		}
	}
	return slog.New(slog.NewJSONHandler(os.Stdout, &slog.HandlerOptions{Level: level}))
}

// WithCorrelationID returns a context carrying the given correlation ID.
func WithCorrelationID(ctx context.Context, id string) context.Context {
	return context.WithValue(ctx, correlationIDKey, id)
}

// CorrelationIDFromContext extracts the correlation ID from context.
func CorrelationIDFromContext(ctx context.Context) string {
	if v, ok := ctx.Value(correlationIDKey).(string); ok {
		return v
	}
	return ""
}

// Info logs at INFO level, adding correlationId if present in ctx.
func Info(ctx context.Context, msg string, args ...any) {
	if id := CorrelationIDFromContext(ctx); id != "" {
		args = append(args, "correlationId", id)
	}
	defaultLogger.InfoContext(ctx, msg, args...)
}

// Warn logs at WARN level.
func Warn(ctx context.Context, msg string, args ...any) {
	if id := CorrelationIDFromContext(ctx); id != "" {
		args = append(args, "correlationId", id)
	}
	defaultLogger.WarnContext(ctx, msg, args...)
}

// Error logs at ERROR level.
func Error(ctx context.Context, msg string, args ...any) {
	if id := CorrelationIDFromContext(ctx); id != "" {
		args = append(args, "correlationId", id)
	}
	defaultLogger.ErrorContext(ctx, msg, args...)
}

// Debug logs at DEBUG level.
func Debug(ctx context.Context, msg string, args ...any) {
	if id := CorrelationIDFromContext(ctx); id != "" {
		args = append(args, "correlationId", id)
	}
	defaultLogger.DebugContext(ctx, msg, args...)
}
